### DISCUSSION TITLE
Action::$table must not be accessed before initialization

### DISCUSSION THREAD

**Discussion by AhmedAbouelkher** [filamentphp/filament#2923] [Help] [answered]
When I try to [customize the creation process](https://filamentphp.com/docs/2.x/admin/resources/relation-managers#customizing-data-before-saving-1), I get promoted with this Laravel error.

```
Typed property Filament\Tables\Actions\Action::$table must not be accessed before initialization
```

## My code
> From ```php public static function table(Table $table): Table``` method
```php
return $table
    ->columns([
        Tables\Columns\TextColumn::make('position')
            ->label('#'),
    ])
    ->filters([
        //
    ])
    ->headerActions([
        CreateAction::make()
            ->process(function (array $data): Model {
                return static::getModel()::create($data);
            })
    ])
    ->actions([
        Tables\Actions\ViewAction::make(),
        Tables\Actions\EditAction::make(),
        Tables\Actions\DeleteAction::make(),
    ])
    ->bulkActions([
        Tables\Actions\DeleteBulkAction::make(),
    ]);
```

## Steps to reproduce:
* Generate a new filament resource (ex: PostsResource with a relationship called `comments`).
* Generate a new Relation manager using this command:
```php artisan make:filament-relation-manager PostsResource comments content```
* Add `CreateAction` to table `headerActions` with a custom process.

## System Information:-
OS: macOS *12.2.1*
Package *filament/filament* with version *2.13*
Laravel Version *9.12.2*
PHP Version *8.1.7*

## Error Logs:
![Screen Shot 2022-06-25 at 9 37 20 PM](https://user-images.githubusercontent.com/56558577/175788195-858d0963-f35c-4d34-b8b7-450c8891025c.png)
![Screen Shot 2022-06-25 at 9 37 11 PM](https://user-images.githubusercontent.com/56558577/175788198-58ecf756-b009-4b8a-93ff-0b457efc0927.png)

**Comment by studiowizjo** [comment]
It is quite common problem when you are trying top customize some things and I also came accross this many times. Of course I don`t remember now exact cases, but luckily got one today. I tried to customize header on CreateRecord page:

**CreateOffer.php:**
```php
protected function getHeader(): View
{
    return view('filament.resources.offers.create-record-header', [
        'heading' => static::$title,
        'actions' => $this->getActions(),
    ]);
}
```

**create-record-header.blade.php**
```blade
<header class="space-y-2 items-start justify-between sm:flex sm:space-y-0 sm:space-x-4 sm:rtl:space-x-reverse sm:py-4 filament-header">
    <x-filament::header.heading>
        {{ $heading }}

        @if ($this->lead->isAlreadyOurCustomer() || true)
            <span class="bg-primary-500/20 dark:text-primary-500 font-medium h- px-2 py-0.5 rounded-xl text-primary-700 text-xs tracking-tight">
                Nasz klient
            </span>
        @endif
    </x-filament::header.heading>

    <x-filament::pages.actions :actions="$actions" class="shrink-0" />
</header>
```

and got quite similiar error: `Typed property Filament\Pages\Actions\Action::$livewire must not be accessed before initialization`.

After some tracking down things I fount that type properties in Livewire should have default values. In my case helped changing:

**vendor\filament\filament\src\Pages\Actions\Concerns\BelongsToLivewire.php**
```php
public function getLivewire(): Page
{
    //was: return $this->livewire;
    return $this->livewire ?? new Page();
}
```
so I think in yours should help:

**vendor\filament\tables\src\Actions\Concerns\BelongsToTable.php**
```php
public function getLivewire(): HasTable
{
    return $this->getTable()->getLivewire() ?? new Table();
}
```

Of course it is topic for PR and it would be worth to track down all similiar places. In some of them I`m not sure what would be the best as they implements interfaces, eg:

**vendor\filament\forms\src\Concerns\BelongsToLivewire.php**
```php
namespace Filament\Forms\Concerns;

use Filament\Forms\Contracts\HasForms;

trait BelongsToLivewire
{
    protected HasForms $livewire;

    public function livewire(HasForms $livewire): static
    {
        $this->livewire = $livewire;

        return $this;
    }

    public function getLivewire(): HasForms
    {
        return $this->livewire;
    }
}
```

So, what with this- use reflections to make default objects? I could make PR but this could be hard for me to test. 

@danharrin what do you think?

**Reply by danharrin** [reply]
So let me explain why this happens so much, why it is actually a useful error that should not be hidden, and how to fix your issue.

In Filament, we use alot of different types of objects. Like form components, table columns, table filters, actions, etc. Each of these objects is able to access data like `$livewire` (the current Livewire instance), `$table` for columns, filters, and actions (the current Table instance), etc.

Let's take your example of a page action:

```php
protected function getActions(): array
{
    return [
        Action::make('runLivewireMethod')->action(fn ($livewire) => $livewire),
    ];
}
```

In this example, we make a new Action object. We then pass in a closure to `$action`, which requires access to the `$livewire` property.

But how do we actually get the Livewire object passed into the action object like that, so the closure can be run?

This actually happens during a process known as "caching". Internally, next to `getActions()`, there are two others. `cacheActions()` and `getCachedActions()`:

```php
protected function getCachedActions(): array
{
    if ($this->cachedActions === null) {
        $this->cacheActions();
    }

    return $this->cachedActions;
}

protected function cacheActions(): void
{
    $this->cachedActions = array_map(
        fn (Action $action): Action => $action->livewire($this)),
        $this->getActions(),
    );
}
```

As you can see, when `getCachedActions()` is called, each Action object from `getActions()` is passed the Livewire object, `$this`. Now, all these action objects have access to `$livewire`, whereas they did not before.

*Filament files never call `getActions()` directly, as this will return Action objects that do not have `$livewire` passed to them yet.* Instead, `getCachedActions()` is always called, which ensures that the correct dependencies have already been injected.

## The error

`Typed property Filament\Pages\Actions\Action::$livewire must not be accessed before initialization` is a useful error. It tells you that $livewire has not been passed into the action yet.

You attempted to suppress the error using `??`. This is not correct, as `$livewire` has still not been passed, and it will result in other errors later.

## Fixing your code

Instead of calling `getActions()` directly, use `getCachedActions()`.

Or even better, don't pass `actions` into the view, just call the method directly:

```blade
<x-filament::pages.actions :actions="$this->getCachedActions()" class="shrink-0" />
```

**Reply by studiowizjo** [reply]
Thank you for comprehensive explanation. You are right, this works and that makes sense. So caching in this case is not only for performance purposes, but also for having access to instance when it is not yet ready?

I don`t know exact lifecycle, but doesn't it work in my suggestion that when we want to reach for an object that we do not have access to, we initialize the default one, but later, when it becomes available, it will not fallback, but takes correct one? If so, then maybe caching methods would not be necessary for this kind of use.

I couldn`t for know observe any downsides of this.

**Reply by danharrin** [reply]
> So caching in this case is not only for performance purposes, but also for having access to instance when it is not yet ready?

Yes, caching is primarily to ensure that the action does not get initialised more than once, for performance. But this was also the most sensible place to inject dependencies.

> I don`t know exact lifecycle, but doesn't it work in my suggestion that when we want to reach for an object that we do not have access to, we initialize the default one, but later, when it becomes available, it will not fallback, but takes correct one? If so, then maybe caching methods would not be necessary for this kind of use.

No, because it will just lead to people getting confused about why their actions don't work, when the actual error is hidden from them. I'd rather be strict and force everyone to use Filament correctly.

**Reply by studiowizjo** [reply]
I got actions on the same page and they work perfectly fine:

```php
    public function getActions() : array
    {
        return [
            Action::make('edit_lead')
                ->color('secondary')
                ->url(fn () => route('filament.resources.leads.edit', $this->lead)),

            Action::make('mark_not_interested')
                ->action(function () {
                    $this->lead->markNotInterested();

                    $nextLead = $this->lead->getNextWaiting();

                    $route = $nextLead
                        ? route('filament.resources.offers.create', $nextLead)
                        : route('filament.resources.leads.index');

                    return redirect($route);
                }),
        ];
    }
```

...but ok, maybe I am missing something which can lead to problems somewhere else.

**Reply by danharrin** [reply]
Yeah but neither of your actions have dependencies... they don't use `$livewire`

**Reply by studiowizjo** [reply]
Thanks for explanations  =).

**Comment by danharrin** [comment, marked as answer]
Actually this was just a typo I made in the docs, the method is called ->using() not ->process(), I've fixed the docs.

**Reply by studiowizjo** [reply]
But the problem stays in other situations - eg. the case which I described above.

**Reply by danharrin** [reply]
The case you described above is completely different to this, I will explain in a reply to that message.
